zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/RestoreFromBackupTool.java

```diff
@@ -377,6 +416,7 @@ public void run() throws IOException {
         copyBackupFilesToLocalTempDir(restoreTempSnapLog);
         processCopiedBackupFiles(restoreTempSnapLog, zxidToRestore);
         copyProcessedRestoredFilesToDestination(restoreTempSnapLog);
+        performSpotRestorationIfConfigured();
```

Comment:
Do we always do `copyProcessedRestoredFilesToDestination()` even when spot restoration is enabled?

Reply:
Yes, because destination dir will be where we store the restored files to be used spot restoration. The temp dir here is just somewhere we temporarily hold the files for processing. But your comment actually reminds me that I need to delete destination directories after the spot restoration is done. Let me add that logic.

Reply:
I see - so the destination dir isn't the snap/dataDir but still a temporary staging path?

Reply:
The destination dir is the snap/dataDir, but yes it is a temporary staging path for spot restoration. Because our expected output for offline restoration is a directory containing restored files, while for spot restoration we don't need the restored files to be left there after the zk server is modified.
Basically when we do spot restoration, we add some extra steps on top of offline restoration. Copy files from backup storage to local temp dir (offline restoration) -> Process the copied files such as truncation (offline restoration) -> Copy processed restored files to local destination (offline restoration) -> Build a zk data tree using files inside local destination (spot restoration) -> Modify zk server (spot restoration) -> Delete local destination directories (spot restoration).

Reply:
Thanks for the detailed, easy-to-understand explanation. Using the steps you outlined, my concern was this step: "Copy processed restored files to local destination (offline restoration)". 

In the case of offline restoration, this makes sense; however, for spot restoration, we do not want to touch what's already in snap/dataDir because the server is live and we don't want anything other than the server itself to write to that path. So naturally, what you want to do is to place the copied-and-processed restored files in a _different_ path other than snap/dataDir and build an in-memory data tree there and spin up a local ZkClient to perform necessary spot writes. But from what you described, we seem to be using the snap/dataDir. This seems a little risky?

Reply:
Discussed with @narendly offline. To make things clearer, the files won't be copied to snap/dataDir if we are doing spot restoration, they will be copied to the local temp dir instead.
Changing the logical flow to:
Copy files from backup storage to local temp dir (offline restoration) -> Process the copied files such as truncation (offline restoration) 
If doing offline restoration --> Copy processed restored files to local destination (offline restoration) 
If doing spot restoration --> Build a zk data tree using files inside local destination (spot restoration) -> Modify zk server (spot restoration) -> Delete local destination directories (spot restoration).